src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -160,6 +170,81 @@ private ImmutableCollections() { }
         }
     }
 
+    /**
+     * Creates a new List from an untrusted array, creating a new array for internal
+     * storage, and checking for and rejecting null elements.
+     *
+     * @param <E> the List's element type
+     * @param input the input array
+     * @return the new list
+     */
+    @SafeVarargs
+    static <E> List<E> listFromArray(E... input) {
+        // copy and check manually to avoid TOCTOU
+        @SuppressWarnings("unchecked")
+        E[] tmp = (E[])new Object[input.length]; // implicit nullcheck of input
+        for (int i = 0; i < input.length; i++) {
+            tmp[i] = Objects.requireNonNull(input[i]);
+        }
+        return new ListN<>(tmp);
+    }
+
+    /**
+     * Creates a new List from a trusted array, checking for and rejecting null
+     * elements.
+     *
+     * <p>A trusted array has no references retained by the caller. It can therefore be
+     * safely reused as the List's internal storage, avoiding a defensive copy. Declared
+     * with Object... instead of E... as the parameter type so that varargs calls don't
+     * accidentally create an array of type other than Object[].
+     *
+     * @param <E> the List's element type
+     * @param input the input array
+     * @return the new list
+     */
+    @SuppressWarnings("unchecked")
+    static <E> List<E> listFromTrustedArray(Object... input) {
+        for (Object o : input) {
+            Objects.requireNonNull(o);
+        }
+
+        switch (input.length) { // implicit null check of elements
+            case 0:
+                return (List<E>) ImmutableCollections.EMPTY_LIST;
+            case 1:
+                return (List<E>) new List12<>(input[0]);
+            case 2:
+                return (List<E>) new List12<>(input[0], input[1]);
+            default:
+                return (List<E>) new ListN<>(input);
+        }
+    }
+
+    /**
+     * Creates a new List from a trusted array, allowing null elements.
+     *
+     * <p>A trusted array has no references retained by the caller. It can therefore be
+     * safely reused as the List's internal storage, avoiding a defensive copy. Declared
+     * with Object... instead of E... as the parameter type so that varargs calls don't
+     * accidentally create an array of type other than Object[].
+     *
+     * <p>Avoids creating a List12 instance, as it cannot accommodate null elements.
+     *
+     * @param <E> the List's element type
+     * @param input the input array
+     * @return the new list
+     */
+    @SuppressWarnings("unchecked")
+    static <E> List<E> listFromTrustedArrayNullsAllowed(Object... input) {
+        if (input.length == 0) {
+            return (List<E>) EMPTY_LIST;
```

Comment:
If we return a `ListN` here, does this mean that `Stream.of().toList().contains(null)` will throw an NPE? But this is incorrect because `toList()` returns a null-tolerant List.

Reply:
Yes, good point, we might need to have a null-tolerant empty list.